packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -304,9 +291,63 @@ export class OdspDocumentService implements IDocumentService {
         });
     }
 
+    private async refreshSessionPeriodically(
+        requestWebsocketTokenFromJoinSession: boolean,
+        response: ISocketStorageDiscovery,
+    ) {
+        for (;;) {
+            let refreshSessionDurationSeconds = response.refreshSessionDurationSeconds;
+            if (refreshSessionDurationSeconds !== undefined) {
+                // 30 seconds is buffer time to refresh the session.
+                let retryAfter = (refreshSessionDurationSeconds * 1000) - 30000;
+                // If retryAfter is negative, try immediately.
+                if (retryAfter < 0) {
```

Comment:
I'm not sure retry immediately is good. I think we should put a min floor (let's say 30 seconds) and if value is below it, do nothing - socket will be disconnected and we will reconnect, but that will ensure that we do not have any tight loops in our code. I.e. I'd worry more about bugs that result in really bad outcomes, versus 0.1% case where we might be slightly more efficient (by not using old flow of reconnects on token expiration)

Reply:
hmm, sounds good. Added return if negative.